File: php/tests/VectorData.php
<?php

declare(strict_types=1);

namespace ArgonGuard\Passwords\Tests;

use ArgonGuard\Passwords\ArgonGuardProfile;
use ArgonGuard\Passwords\InvalidInputException;
use ArgonGuard\Passwords\MalformedHashException;
use ArgonGuard\Passwords\PolicyViolationException;
use ArgonGuard\Passwords\UnsupportedAlgorithmException;
use ArgonGuard\Passwords\UnsupportedEnvironmentException;

/** 凍結向量與規格 artifact 載入 helper（權威來源：spec/）。 */
final class VectorData
{
    public const SPEC_DIR = __DIR__ . '/../../spec';

    private function __construct()
    {
    }

    /** @return list<array<string, mixed>> */
    public static function entries(string $file): array
    {
        $json = self::loadJson(self::SPEC_DIR . '/vectors/v1/' . $file);

        return $json['entries'];
    }

    /** @return array<string, mixed> */
    public static function entry(string $file, string $id): array
    {
        foreach (self::entries($file) as $entry) {
            if ($entry['id'] === $id) {
                return $entry;
            }
        }

        throw new \RuntimeException("{$file}: {$id} not found");
    }

    /** @return array<string, mixed> */
    public static function engineUnits(): array
    {
        return self::loadJson(self::SPEC_DIR . '/engine-units.json');
    }

    /** @return array<string, mixed> */
    public static function reasonCodes(): array
    {
        return self::loadJson(self::SPEC_DIR . '/reason-codes.json');
    }

    /** @return class-string<\ArgonGuard\Passwords\ArgonGuardException> */
    public static function errorClass(string $name): string
    {
        return match ($name) {
            'MalformedHash' => MalformedHashException::class,
            'UnsupportedAlgorithm' => UnsupportedAlgorithmException::class,
            'PolicyViolation' => PolicyViolationException::class,
            'InvalidInput' => InvalidInputException::class,
            'UnsupportedEnvironment' => UnsupportedEnvironmentException::class,
            default => throw new \RuntimeException("unknown error category: {$name}"),
        };
    }

    public static function profile(string $name): ArgonGuardProfile
    {
        return match ($name) {
            'default' => ArgonGuardProfile::Default,
            'high' => ArgonGuardProfile::High,
            'highest' => ArgonGuardProfile::Highest,
            default => throw new \RuntimeException("unknown profile: {$name}"),
        };
    }

    public static function hex(string $hex): string
    {
        return $hex === '' ? '' : (string) hex2bin($hex);
    }

    /** @return array<string, mixed> */
    private static function loadJson(string $path): array
    {
        return json_decode((string) file_get_contents($path), true, 64, JSON_THROW_ON_ERROR);
    }
}

File: dotnet/tests/ArgonGuard.Passwords.Tests/VectorData.cs
using System;
using System.IO;
using System.Text.Json;

namespace ArgonGuard.Passwords.Tests;

/// <summary>凍結向量載入器：由測試輸出目錄向上尋找 repo 根的 spec/vectors/v1。</summary>
internal static class VectorData
{
    private static readonly Lazy<string> VectorDir = new(() =>
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir is not null)
        {
            var candidate = Path.Combine(dir.FullName, "spec", "vectors", "v1");
            if (Directory.Exists(candidate)) return candidate;
            dir = dir.Parent;
        }
        throw new InvalidOperationException("spec/vectors/v1 not found above test directory");
    });

    public static JsonElement[] Entries(string fileName)
    {
        using var doc = JsonDocument.Parse(File.ReadAllText(Path.Combine(VectorDir.Value, fileName)));
        var entries = doc.RootElement.GetProperty("entries");
        var result = new JsonElement[entries.GetArrayLength()];
        int i = 0;
        foreach (var e in entries.EnumerateArray()) result[i++] = e.Clone();
        return result;
    }

    public static string SpecFile(string relative)
        => Path.Combine(VectorDir.Value, "..", "..", relative);

    public static byte[] Hex(string hex)
    {
        var bytes = new byte[hex.Length / 2];
        for (int i = 0; i < bytes.Length; i++)
            bytes[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
        return bytes;
    }

    /// <summary>bytes → .NET string（測試中重建字串型 API 輸入；向量皆為合法 UTF-8）。</summary>
    public static string Utf8(byte[] bytes) => System.Text.Encoding.UTF8.GetString(bytes);

    public static ArgonGuardProfile Profile(string name) => name switch
    {
        "default" => ArgonGuardProfile.Default,
        "high" => ArgonGuardProfile.High,
        "highest" => ArgonGuardProfile.Highest,
        _ => throw new ArgumentOutOfRangeException(nameof(name), name, null),
    };

    public static Type ErrorType(string category) => category switch
    {
        "MalformedHash" => typeof(MalformedHashException),
        "UnsupportedAlgorithm" => typeof(UnsupportedAlgorithmException),
        "PolicyViolation" => typeof(PolicyViolationException),
        "InvalidInput" => typeof(InvalidInputException),
        "UnsupportedEnvironment" => typeof(UnsupportedEnvironmentException),
        _ => throw new ArgumentOutOfRangeException(nameof(category), category, null),
    };
}

/// <summary>needs-rehash 向量中 legacyRegistered=true 時註冊的 bcrypt 前綴認領器（不做真驗證）。</summary>
internal sealed class FakeBcryptVerifier : ILegacyPasswordVerifier
{
    public bool CanHandle(string encodedHash) => encodedHash.StartsWith("$2b$", StringComparison.Ordinal);
    public bool Verify(string password, string encodedHash) => false;
}
